File: backend/models/changelist.go
package models

import "time"

// Constants for changelist types
const (
	ChangelistTypeCustom    = "custom"
	ChangelistTypeTracked   = "tracked"
	ChangelistTypeUntracked = "untracked"
)

// Validation constants
const (
	MaximumGroupNameLength = 100
	MinimumGroupNameLength = 1
)

// Schema version for future migrations
const ChangelistSchemaVersion = 1

// Changelist represents a group of file changes
type Changelist struct {
	IdentifierValue    string           `json:"id"`
	GroupName          string           `json:"name"`
	GroupType          string           `json:"type"` // "custom", "tracked", "untracked"
	CreatedAtTimestamp time.Time        `json:"createdAt"`
	UpdatedAtTimestamp time.Time        `json:"updatedAt"`
	FileItems          []ChangelistItem `json:"items"`
	IsSystemGenerated  bool             `json:"isSystemGenerated"`      // true for tracked/untracked
	OrderIndex         int              `json:"orderIndex"`             // for sorting groups
	ColorHexCode       string           `json:"colorHexCode,omitempty"` // optional UI color
	DescriptionText    string           `json:"description,omitempty"`  // optional description
}

// ChangelistItem represents a single file within a changelist group
type ChangelistItem struct {
	FilePath                 string    `json:"path"`
	TrackedSnapshotHash      string    `json:"trackedSnapshotHash,omitempty"`      // git hash when added (optional)
	NotesText                string    `json:"notes,omitempty"`                    // user notes about this file
	IsMissingFromWorkingTree bool      `json:"isMissingFromWorkingTree,omitempty"` // true if file was deleted
	AddedAtTimestamp         time.Time `json:"addedAt"`
	LastModifiedTimestamp    time.Time `json:"lastModifiedAt"`
}

// ChangelistConfiguration represents the root configuration file structure
type ChangelistConfiguration struct {
	SchemaVersion  int          `json:"version"`
	CustomGroups   []Changelist `json:"groups"`
	LastSavedAt    time.Time    `json:"lastSavedAt"`
	RepositoryPath string       `json:"repositoryPath"`
}

// ArchiveMetadata represents metadata for an archived changelist
type ArchiveMetadata struct {
	ArchiveIdentifier   string    `json:"id"`
	OriginalGroupName   string    `json:"originalGroupName"`
	ArchiveName         string    `json:"archiveName"`
	CreatedAtTimestamp  time.Time `json:"createdAt"`
	ArchivedAtTimestamp time.Time `json:"archivedAt"`
	FilePathsList       []string  `json:"filePaths"`
	OriginalBranchName  string    `json:"originalBranch"`
	OriginalCommitHash  string    `json:"originalCommitHash"`
	DescriptionText     string    `json:"description,omitempty"`
	TagsList            []string  `json:"tags,omitempty"`
	DiffFileSizeBytes   int64     `json:"diffFileSize"`
	TotalFilesCount     int       `json:"totalFilesCount"`
	TotalAdditionsCount int       `json:"totalAdditions"`
	TotalDeletionsCount int       `json:"totalDeletions"`
}

// ArchiveListEntry represents a summary of an archive for list views
type ArchiveListEntry struct {
	ArchiveIdentifier   string    `json:"id"`
	ArchiveName         string    `json:"archiveName"`
	OriginalGroupName   string    `json:"originalGroupName"`
	ArchivedAtTimestamp time.Time `json:"archivedAt"`
	TotalFilesCount     int       `json:"totalFilesCount"`
	DiffFileSizeBytes   int64     `json:"diffFileSize"`
	TagsList            []string  `json:"tags,omitempty"`
}

// CreateChangelistRequest represents a request to create a new changelist
type CreateChangelistRequest struct {
	GroupName       string   `json:"name"`
	DescriptionText string   `json:"description,omitempty"`
	ColorHexCode    string   `json:"colorHexCode,omitempty"`
	InitialFiles    []string `json:"initialFiles,omitempty"`
}

// UpdateChangelistRequest represents a request to update an existing changelist
type UpdateChangelistRequest struct {
	IdentifierValue string `json:"id"`
	GroupName       string `json:"name,omitempty"`
	DescriptionText string `json:"description,omitempty"`
	ColorHexCode    string `json:"colorHexCode,omitempty"`
}

// MoveFilesRequest represents a request to move files between changelists
type MoveFilesRequest struct {
	SourceGroupIdentifier      string   `json:"sourceGroupId"`
	DestinationGroupIdentifier string   `json:"destinationGroupId"`
	FilePathsList              []string `json:"filePaths"`
}

// CreateArchiveRequest represents a request to archive a changelist
type CreateArchiveRequest struct {
	GroupIdentifier string   `json:"groupId"`
	ArchiveName     string   `json:"archiveName"`
	DescriptionText string   `json:"description,omitempty"`
	TagsList        []string `json:"tags,omitempty"`
}

// RestoreArchiveRequest represents a request to restore an archived changelist
type RestoreArchiveRequest struct {
	ArchiveIdentifier     string `json:"archiveId"`
	TargetGroupIdentifier string `json:"targetGroupId,omitempty"` // optional, creates new group if empty
	NewGroupName          string `json:"newGroupName,omitempty"`  // used if creating new group
}

File: frontend/src/types/changelist.ts
// Constants for changelist types
export const CHANGELIST_TYPE_CUSTOM = 'custom';
export const CHANGELIST_TYPE_TRACKED = 'tracked';
export const CHANGELIST_TYPE_UNTRACKED = 'untracked';

export type ChangelistType =
  | typeof CHANGELIST_TYPE_CUSTOM
  | typeof CHANGELIST_TYPE_TRACKED
  | typeof CHANGELIST_TYPE_UNTRACKED;

// Validation constants
export const MAXIMUM_GROUP_NAME_LENGTH = 100;
export const MINIMUM_GROUP_NAME_LENGTH = 1;

// Schema version for future migrations
export const CHANGELIST_SCHEMA_VERSION = 1;

// Changelist represents a group of file changes
export interface Changelist {
  id: string;
  name: string;
  type: ChangelistType;
  createdAt: string; // ISO 8601 timestamp
  updatedAt: string; // ISO 8601 timestamp
  items: ChangelistItem[];
  isSystemGenerated: boolean; // true for tracked/untracked
  orderIndex: number; // for sorting groups
  colorHexCode?: string; // optional UI color
  description?: string; // optional description
}

// ChangelistItem represents a single file within a changelist group
export interface ChangelistItem {
  path: string;
  trackedSnapshotHash?: string; // git hash when added (optional)
  notes?: string; // user notes about this file
  isMissingFromWorkingTree?: boolean; // true if file was deleted
  addedAt: string; // ISO 8601 timestamp
  lastModifiedAt: string; // ISO 8601 timestamp
}

// ChangelistConfiguration represents the root configuration file structure
export interface ChangelistConfiguration {
  version: number;
  groups: Changelist[];
  lastSavedAt: string; // ISO 8601 timestamp
  repositoryPath: string;
}

// ArchiveMetadata represents metadata for an archived changelist
export interface ArchiveMetadata {
  id: string;
  originalGroupName: string;
  archiveName: string;
  createdAt: string; // ISO 8601 timestamp
  archivedAt: string; // ISO 8601 timestamp
  filePaths: string[];
  originalBranch: string;
  originalCommitHash: string;
  description?: string;
  tags?: string[];
  diffFileSize: number;
  totalFilesCount: number;
  totalAdditions: number;
  totalDeletions: number;
}

// ArchiveListEntry represents a summary of an archive for list views
export interface ArchiveListEntry {
  id: string;
  archiveName: string;
  originalGroupName: string;
  archivedAt: string; // ISO 8601 timestamp
  totalFilesCount: number;
  diffFileSize: number;
  tags?: string[];
}

// CreateChangelistRequest represents a request to create a new changelist
export interface CreateChangelistRequest {
  name: string;
  description?: string;
  colorHexCode?: string;
  initialFiles?: string[];
}

// UpdateChangelistRequest represents a request to update an existing changelist
export interface UpdateChangelistRequest {
  id: string;
  name?: string;
  description?: string;
  colorHexCode?: string;
}

// MoveFilesRequest represents a request to move files between changelists
export interface MoveFilesRequest {
  sourceGroupId: string;
  destinationGroupId: string;
  filePaths: string[];
}

// CreateArchiveRequest represents a request to archive a changelist
export interface CreateArchiveRequest {
  groupId: string;
  archiveName: string;
  description?: string;
  tags?: string[];
}

// RestoreArchiveRequest represents a request to restore an archived changelist
export interface RestoreArchiveRequest {
  archiveId: string;
  targetGroupId?: string; // optional, creates new group if empty
  newGroupName?: string; // used if creating new group
}

// Helper type for UI state
export interface ChangelistUIState {
  selectedGroupId: string | null;
  selectedFilePaths: string[];
  expandedGroupIds: string[];
  isLoading: boolean;
  error: string | null;
}

// Helper type for archive UI state
export interface ArchiveUIState {
  selectedArchiveId: string | null;
  isLoading: boolean;
  error: string | null;
}
